### DISCUSSION TITLE
best practice for local, development, staging tables?

### DISCUSSION THREAD

**Discussion by jsjoeio** [supabase/supabase#2590] [Questions] [answered]
What's the best practice for having multiple tables of the same entity? 

i.e. a `users` table.

Do I create one for each environment and name them like so:
- `users - production`
- `users - development`
- `users - staging`

Wondering what the best approach is here.

**Comment by silentworks** [comment, marked as answer]
My approach would be to create individual projects for each of these environments. I would probably only use this for staging and production though since you can use local supabase for development environment. Use some sort of migration tool like [ley](https://github.com/lukeed/ley) to create initial table migrations or you can use the supabase cli to create diff migrations between environments.

Another approach would be to use separate schemas inside the postgres database, the main issue with this is that all environments would share the `auth, storage` schemas. So you wouldn't be able to have the same email address with different passwords for the different environments. You also wouldn't be able to upload the same file to storage in different environments.

**Reply by jsjoeio** [reply]
> My approach would be to create individual projects for each of these environments.

Ah, that makes sense.

> since you can use local supabase for development environment

Oh I didn't know that. TIL. I'm not even sure how to do this, I guess I'll have to take a closer look at the docs. 

Thanks for the explanation!

**Reply by fergusmeiklejohn** [reply]
local development is actually much better and easier than I thought it would be. Coming from firestore I expected to need to use Supabase's CLI a lot to use it locally, but I used to setup and haven't used it again (apart from starting and stopping the instance) It's just a docker container with the database and all the rest, so I use Datagrip and do almost everything on my machine. Datagrip is brilliant. I have no experience with SQL but I'm writing complicated queries, I can export the data returned to JSON, copy it into my project, write a function to get to get the data into the shape I need, go back to datagrip, create a View and query it from the client. You can also connect to the production Supabase instance in the cloud with datagrip. I'm now looking at how to migrate changes since supabase CLI doesn't really provide that yet.

**Reply by sandypockets** [reply]
> local development is actually much better and easier than I thought it would be. Coming from firestore I expected to need to use Supabase's CLI a lot to use it locally, but I used to setup and haven't used it again (apart from starting and stopping the instance) It's just a docker container with the database and all the rest, so I use Datagrip and do almost everything on my machine. Datagrip is brilliant. I have no experience with SQL but I'm writing complicated queries, I can export the data returned to JSON, copy it into my project, write a function to get to get the data into the shape I need, go back to datagrip, create a View and query it from the client. You can also connect to the production Supabase instance in the cloud with datagrip. I'm now looking at how to migrate changes since supabase CLI doesn't really provide that yet.

@fergusmeiklejohn I'm running into some challenges getting things up and running in DataGrip while running Supabase locally. Did you run into any errors like that? I'm getting an error that 'relation "public.users" does not exist', even though it definitely does in DataGrip.

**Reply by fergusmeiklejohn** [reply]
I've had no trouble with datagrip, except that I needed to create auth.users otherwise foreign keys in the public schema  that reference auth.users will be dropped. The auth schema isn't included in schemas by default when connecting in datagrip.
I'm on an M1 Mac now so I've stopped using the local docker supabase, because of a Postgrest bug. And I've found makes no difference to work against a live dev DB.
I'm also using the webstorm datagrip plugin which is cheaper and it's so useful to have everything in one IDE.
